File: api/chunithm/music.py
from typing import Tuple, List
from sqlalchemy import select, desc, and_
from fastapi_cache.decorator import cache
from fastapi_limiter.depends import RateLimiter
from fastapi import APIRouter, Request, HTTPException, Depends, Query
from fastapi.responses import ORJSONResponse

from modules.exceptions import APIException
from utils import chunithm_music_engine as engine, parse_json_body
from modules.schemas.chunithm import (
    MusicInfoSchema,
    MusicDifficultySchema,
    MusicBatchItemSchema,
    MusicBatchRequestSchema,
    ChartDataSchema,
    AllMusicResponse,
    MusicBatchResultSchema, MusicDifficultyResponse, MusicInfoResponse, ChartDataResponse
)
from modules.sql.tables.chunithm import ChunithmMusicDifficulty, ChunithmMusic, ChunithmChartData

music_api = APIRouter(prefix="/music", tags=["music_api"])


@music_api.get("/all-music",
               response_model=AllMusicResponse,
               summary="查询所有乐曲",
               description="查询数据库中所有乐曲信息",
               dependencies=[Depends(RateLimiter(times=10, seconds=5))])
@cache(expire=300)
async def get_all_music() -> AllMusicResponse:
    async with engine.session() as session:
        stmt = select(ChunithmMusic)
        result = await session.execute(stmt)
        rows = result.scalars().all()
        return AllMusicResponse(data=[MusicInfoSchema.model_validate(row) for row in rows])


@music_api.get("/{music_id}/difficulty-info",
               response_model=MusicDifficultyResponse,
               summary="查询乐曲定数",
               description="根据提供的乐曲ID，返回对应乐曲基本信息",
               dependencies=[Depends(RateLimiter(times=10, seconds=1))])
@cache(expire=300)
async def get_music_difficulty_info(
    music_id: int, version: str = Query(..., description="查询版本，如2.30")
) -> MusicDifficultyResponse:
    record = await engine.select(
        ChunithmMusicDifficulty,
        and_(ChunithmMusicDifficulty.music_id == music_id, ChunithmMusicDifficulty.version == version),
        one_result=True,
    )
    if record:
        return MusicDifficultyResponse.model_validate(record)
    async with engine.session() as session:
        stmt_latest = (
            select(ChunithmMusicDifficulty)
            .where(ChunithmMusicDifficulty.music_id == music_id)
            .order_by(desc(ChunithmMusicDifficulty.version))
            .limit(1)
        )
        result = await session.execute(stmt_latest)
        latest = result.scalar_one_or_none()
        if latest:
            return MusicDifficultyResponse.model_validate(latest)
        raise APIException(status=404, message="No difficulty data")


@music_api.get("/{music_id}/basic-info",
               response_model=MusicInfoResponse,
               summary="查询乐曲基本信息",
               description="根据提供的乐曲ID，返回对应乐曲基本信息",
               dependencies=[Depends(RateLimiter(times=10, seconds=1))])
@cache(expire=300)
async def get_music_basic_info(music_id: int) -> MusicInfoResponse:
    music = await engine.select(ChunithmMusic, ChunithmMusic.music_id == music_id, one_result=True)
    if not music:
        raise APIException(status=404, message="Music not found")
    return MusicInfoResponse.model_validate(music)


@music_api.get("/{music_id}/chart-data",
               response_model=ChartDataResponse,
               summary="查询乐曲谱面数据",
               description="根据提供的乐曲ID，返回对应乐曲谱面数据",
               dependencies=[Depends(RateLimiter(times=10, seconds=1))])
@cache(expire=300)
async def get_music_chart_data(music_id: int) -> ChartDataResponse:
    chart_rows = await engine.select(ChunithmChartData, ChunithmChartData.music_id == music_id)
    if not chart_rows:
        raise APIException(status=404, message="No chart data found")
    chart_data = [ChartDataSchema.model_validate(row) for row in chart_rows]
    return ChartDataResponse(data=chart_data)


@music_api.post(
    "/query-batch",
    response_model=MusicBatchResultSchema,
    summary="批量查询乐曲信息与定数",
    description="根据提供的乐曲ID列表和版本号，返回对应的乐曲信息与定数",
    dependencies=[Depends(RateLimiter(times=3, seconds=1))],
)
async def get_music_data_batch(
    validated: MusicBatchRequestSchema = Depends(parse_json_body(engine, MusicBatchRequestSchema))
) -> MusicBatchResultSchema:
    music_ids = validated.music_ids
    version = validated.version
    async with engine.session() as session:
        music_stmt = select(ChunithmMusic).where(ChunithmMusic.music_id.in_(music_ids))
        difficulty_stmt = select(ChunithmMusicDifficulty).where(ChunithmMusicDifficulty.music_id.in_(music_ids))
        music_result = await session.execute(music_stmt)
        difficulty_result = await session.execute(difficulty_stmt)
        music_map = {m.music_id: m for m in music_result.scalars().all()}
        difficulty_map: dict[int, MusicDifficultySchema] = {}
        for d in sorted(difficulty_result.scalars().all(), key=lambda x: x.version, reverse=True):
            if d.music_id not in difficulty_map or d.version == version:
                difficulty_map[d.music_id] = MusicDifficultySchema.model_validate(d)
        result: dict[int, MusicBatchItemSchema] = {}
        for mid in music_ids:
            music = music_map.get(mid)
            difficulty = difficulty_map.get(mid)
            difficulty_list = [
                difficulty.diff0_const,
                difficulty.diff1_const,
                difficulty.diff2_const,
                difficulty.diff3_const,
                difficulty.diff4_const,
            ] if difficulty else [None] * 5
            if music:
                info = MusicInfoSchema.model_validate(music)
                version_val = music.version
            else:
                info = MusicInfoSchema(
                    music_id=mid,
                    title="Unknown",
                    artist="Unknown",
                    category="Unknown",
                    version=None,
                    releaseDate=None,
                    isDeleted=False,
                    deletedVersion=None,
                )
                version_val = None
            result[mid] = MusicBatchItemSchema(
                version=version_val,
                difficulty=difficulty_list,
                info=info,
            )
        return MusicBatchResultSchema(__root__=result)

File: api/chunithm/music.go
package chunithm

import (
	"context"
	"haruki-database/api"
	"haruki-database/config"
	"sort"
	"time"

	entchuniMusic "haruki-database/database/schema/chunithm/music"
	"haruki-database/database/schema/chunithm/music/chunithmchartdata"
	"haruki-database/database/schema/chunithm/music/chunithmmusic"
	"haruki-database/database/schema/chunithm/music/chunithmmusicdifficulty"

	"github.com/gofiber/fiber/v3"
	"github.com/redis/go-redis/v9"
)

func (h *MusicHandler) GetAllMusic(c fiber.Ctx) error {
	ctx := context.Background()
	now := time.Now()
	key, cached, hit, err := api.CacheQuery(ctx, c, h.svc.redisClient, CacheNSMusic)
	if err != nil {
		return api.InternalError(c)
	}
	if hit {
		return c.Status(fiber.StatusOK).JSON(cached)
	}
	rows, err := h.svc.client.ChunithmMusic.
		Query().
		Where(
			chunithmmusic.Or(
				chunithmmusic.ReleaseDateLTE(now),
				chunithmmusic.ReleaseDateIsNil(),
			),
		).
		All(ctx)
	if err != nil {
		return api.InternalError(c)
	}
	result := make([]MusicInfoSchema, len(rows))
	for i, row := range rows {
		deleted := row.IsDeleted
		result[i] = MusicInfoSchema{
			MusicID:        row.MusicID,
			Title:          row.Title,
			Artist:         row.Artist,
			Category:       row.Category,
			Version:        row.Version,
			ReleaseDate:    row.ReleaseDate,
			IsDeleted:      &deleted,
			DeletedVersion: row.DeletedVersion,
		}
	}
	return api.CachedJSONResponse(ctx, c, h.svc.redisClient, config.Cfg.Backend.APICacheTTL, key, fiber.StatusOK, "ok", result)
}

func (h *MusicHandler) GetDifficultyInfo(c fiber.Ctx) error {
	ctx := context.Background()
	musicID := fiber.Params[int](c, "music_id", -1)
	if musicID <= 0 {
		return api.JSONResponse(c, fiber.StatusBadRequest, "invalid music_id")
	}
	version := c.Query("version")
	if version == "" {
		return api.JSONResponse(c, fiber.StatusBadRequest, "version required")
	}
	key, cached, hit, err := api.CacheQuery(ctx, c, h.svc.redisClient, CacheNSMusic)
	if err != nil {
		return api.InternalError(c)
	}
	if hit {
		return c.Status(fiber.StatusOK).JSON(cached)
	}
	record, _ := h.svc.client.ChunithmMusicDifficulty.
		Query().
		Where(chunithmmusicdifficulty.MusicIDEQ(musicID), chunithmmusicdifficulty.VersionEQ(version)).
		First(ctx)
	if record != nil {
		payload := MusicDifficultySchema{
			MusicID: record.MusicID,
			Version: record.Version,
			Diff0:   record.Diff0Const,
			Diff1:   record.Diff1Const,
			Diff2:   record.Diff2Const,
			Diff3:   record.Diff3Const,
			Diff4:   record.Diff4Const,
		}
		return api.CachedJSONResponse(ctx, c, h.svc.redisClient, config.Cfg.Backend.APICacheTTL, key, fiber.StatusOK, "ok", payload)
	}
	latest, _ := h.svc.client.ChunithmMusicDifficulty.
		Query().
		Where(chunithmmusicdifficulty.MusicIDEQ(musicID)).
		Order(entchuniMusic.Desc(chunithmmusicdifficulty.FieldVersion)).
		First(ctx)
	if latest != nil {
		payload := MusicDifficultySchema{
			MusicID: latest.MusicID,
			Version: latest.Version,
			Diff0:   latest.Diff0Const,
			Diff1:   latest.Diff1Const,
			Diff2:   latest.Diff2Const,
			Diff3:   latest.Diff3Const,
			Diff4:   latest.Diff4Const,
		}
		return api.CachedJSONResponse(ctx, c, h.svc.redisClient, config.Cfg.Backend.APICacheTTL, key, fiber.StatusOK, "ok", payload)
	}
	return api.JSONResponse(c, fiber.StatusNotFound, "No difficulty data")
}

func (h *MusicHandler) GetBasicInfo(c fiber.Ctx) error {
	ctx := context.Background()
	musicID := fiber.Params[int](c, "music_id", -1)
	if musicID <= 0 {
		return api.JSONResponse(c, fiber.StatusBadRequest, "invalid music_id")
	}
	key, cached, hit, err := api.CacheQuery(ctx, c, h.svc.redisClient, CacheNSMusic)
	if err != nil {
		return api.InternalError(c)
	}
	if hit {
		return c.Status(fiber.StatusOK).JSON(cached)
	}
	row, _ := h.svc.client.ChunithmMusic.
		Query().
		Where(chunithmmusic.MusicIDEQ(musicID)).
		First(ctx)
	if row == nil {
		return api.JSONResponse(c, fiber.StatusNotFound, "Music not found")
	}
	deleted := row.IsDeleted
	return api.CachedJSONResponse(ctx, c, h.svc.redisClient, config.Cfg.Backend.APICacheTTL, key, fiber.StatusOK, "ok", MusicInfoSchema{
		MusicID:        row.MusicID,
		Title:          row.Title,
		Artist:         row.Artist,
		Category:       row.Category,
		Version:        row.Version,
		ReleaseDate:    row.ReleaseDate,
		IsDeleted:      &deleted,
		DeletedVersion: row.DeletedVersion,
	})
}

func (h *MusicHandler) GetChartData(c fiber.Ctx) error {
	ctx := context.Background()
	musicID := fiber.Params[int](c, "music_id", -1)
	if musicID <= 0 {
		return api.JSONResponse(c, fiber.StatusBadRequest, "invalid music_id")
	}
	key, cached, hit, err := api.CacheQuery(ctx, c, h.svc.redisClient, CacheNSMusic)
	if err != nil {
		return api.InternalError(c)
	}
	if hit {
		return c.Status(fiber.StatusOK).JSON(cached)
	}
	rows, _ := h.svc.client.ChunithmChartData.
		Query().
		Where(chunithmchartdata.MusicIDEQ(musicID)).
		All(ctx)
	if len(rows) == 0 {
		return api.JSONResponse(c, fiber.StatusNotFound, "No chart data found")
	}
	result := make([]ChartDataSchema, len(rows))
	for i, r := range rows {
		result[i] = ChartDataSchema{
			Difficulty: r.Difficulty,
			Creator:    r.Creator,
			BPM:        r.Bpm,
			TapCount:   r.TapCount,
			HoldCount:  r.HoldCount,
			SlideCount: r.SlideCount,
			AirCount:   r.AirCount,
			FlickCount: r.FlickCount,
			TotalCount: r.TotalCount,
		}
	}
	return api.CachedJSONResponse(ctx, c, h.svc.redisClient, config.Cfg.Backend.APICacheTTL, key, fiber.StatusOK, "ok", result)
}

func (h *MusicHandler) QueryBatch(c fiber.Ctx) error {
	ctx := context.Background()
	var req struct {
		MusicIDs []int  `json:"music_ids"`
		Version  string `json:"version"`
	}
	if err := c.Bind().Body(&req); err != nil {
		return api.JSONResponse(c, fiber.StatusBadRequest, api.ErrInvalidRequest)
	}
	musicRows, _ := h.svc.client.ChunithmMusic.
		Query().
		Where(chunithmmusic.MusicIDIn(req.MusicIDs...)).
		All(ctx)
	musicMap := make(map[int]*entchuniMusic.ChunithmMusic)
	for _, m := range musicRows {
		musicMap[m.MusicID] = m
	}
	diffRows, _ := h.svc.client.ChunithmMusicDifficulty.
		Query().
		Where(chunithmmusicdifficulty.MusicIDIn(req.MusicIDs...)).
		All(ctx)
	sort.Slice(diffRows, func(i, j int) bool {
		return diffRows[i].Version > diffRows[j].Version
	})
	diffMap := make(map[int]*entchuniMusic.ChunithmMusicDifficulty)
	for _, d := range diffRows {
		if _, ok := diffMap[d.MusicID]; !ok || d.Version == req.Version {
			diffMap[d.MusicID] = d
		}
	}
	result := make(map[int]MusicBatchItemSchema)
	for _, mid := range req.MusicIDs {
		music := musicMap[mid]
		diff := diffMap[mid]
		var diffList []*float64
		if diff != nil {
			diffList = []*float64{
				diff.Diff0Const, diff.Diff1Const, diff.Diff2Const,
				diff.Diff3Const, diff.Diff4Const,
			}
		} else {
			diffList = []*float64{nil, nil, nil, nil, nil}
		}
		var info MusicInfoSchema
		var version *string
		if music != nil {
			deleted := music.IsDeleted
			info = MusicInfoSchema{
				MusicID:        music.MusicID,
				Title:          music.Title,
				Artist:         music.Artist,
				Category:       music.Category,
				Version:        music.Version,
				ReleaseDate:    music.ReleaseDate,
				IsDeleted:      &deleted,
				DeletedVersion: music.DeletedVersion,
			}
			version = music.Version
		} else {
			title := "Unknown"
			artist := "Unknown"
			category := "Unknown"
			isDeleted := false
			info = MusicInfoSchema{
				MusicID:   mid,
				Title:     title,
				Artist:    artist,
				Category:  &category,
				IsDeleted: &isDeleted,
			}
			version = nil
		}
		result[mid] = MusicBatchItemSchema{
			Version:    version,
			Difficulty: diffList,
			Info:       info,
		}
	}
	return api.JSONResponse(c, fiber.StatusOK, "success", result)
}

func registerMusicRoutes(r fiber.Router, client *entchuniMusic.Client, redisClient *redis.Client) {
	svc := NewMusicService(client, redisClient)
	h := NewMusicHandler(svc)
	apiGroup := r.Group("/music")

	apiGroup.Get("/all-music", h.GetAllMusic)
	apiGroup.Get("/:music_id/difficulty-info", h.GetDifficultyInfo)
	apiGroup.Get("/:music_id/basic-info", h.GetBasicInfo)
	apiGroup.Get("/:music_id/chart-data", h.GetChartData)
	apiGroup.Post("/query-batch", h.QueryBatch)
}
